### PR TITLE
update migration script to 15.3

### PR DESCRIPTION
**Pull request by jordimassaguerpla**
## What does this PR change?

update migration script to 15.3

## GUI diff

No difference.

- [X] **DONE**

## Documentation
- No documentation needed

- [X] **DONE**

## Test coverage
- No tests
- [X] **DONE**

## Links

https://github.com/SUSE/spacewalk/issues/15036

- [X] **DONE**

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

**Comment by jordimassaguerpla** [comment]
I just tested this and had this errors

image

**Comment by mcalmer** [comment]
Afaik python2-salt does not exist anymore. We should try to add a weakremover in the release package

**Comment by juliogonzalez** [comment]
@jordimassaguerpla not sure from where your packages are. Are you using Master?

I am asking because Master is already Uyuni 2021.06, and your screenshot still shows Uyuni 2021.05. If that's the case, then the server pattern still has python2-salt, unlike Uyuni 2021.06.

File: susemanager/bin/server-migrator.sh [hunk 1]
```diff
@@ -22,18 +22,18 @@ echo
...
 mkdir /etc/zypp/repos.d
-zypper ar -n "Main Repository" http://download.opensuse.org/distribution/leap/15.2/repo/oss repo-oss
-zypper ar -n "Main Update Repository" http://download.opensuse.org/update/leap/15.2/oss repo-update
-zypper ar -n "Non-OSS Repository" http://download.opensuse.org/distribution/leap/15.2/repo/non-oss repo-non-oss
-zypper ar -n "Update Repository (Non-Oss)" http://download.opensuse.org/update/leap/15.2/non-oss/ repo-update-non-oss
+zypper ar -n "Main Repository" http://download.opensuse.org/distribution/leap/15.3/repo/oss repo-oss
+zypper ar -n "Main Update Repository" http://download.opensuse.org/update/leap/15.3/oss repo-update
+zypper ar -n "Non-OSS Repository" http://download.opensuse.org/distribution/leap/15.3/repo/non-oss repo-non-oss
+zypper ar -n "Update Repository (Non-Oss)" http://download.opensuse.org/update/leap/15.3/non-oss/ repo-update-non-oss
 zypper ar -n "Uyuni Server Stable" https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Stable/images/repo/Uyuni-Server-POOL-x86_64-Media1/ uyuni-server-stable
```
Comment (juliogonzalez) [inline susemanager/bin/server-migrator.sh:29]: This is your problem. For the test to work, this has to be Master.

**Comment by jordimassaguerpla** [comment]
I had to remove python2-salt, postgresql12-llvm and product Uyuni-Server for "zypper dup" to work.
@juliogonzalez @mcalmer How do we usually do this? Should I just add to this script the zypper remove commands or should we make some package have an obsoletes?

**Comment by jordimassaguerpla** [comment]
@juliogonzalez I have not seen your comment above about using master, I will try now.

**Comment by juliogonzalez** [comment]
Make sure you test witthout removing python2-salt and postgresql12-llvm first, just in case it's not really needed.

File: susemanager/bin/server-migrator.sh (hunk 1 above)
Comment (jordimassaguerpla) [reply susemanager/bin/server-migrator.sh:29]: Even with Master, I still need to uninstall python2-salt and postgresql12-llvmjit

**Review by jordimassaguerpla — COMMENTED** [review]
> Afaik python2-salt does not exist anymore. We should try to add a weakremover in the release package

What is "a weakremover in the release package" ?

File: susemanager/bin/server-migrator.sh (hunk 1 above)
Comment (juliogonzalez) [reply susemanager/bin/server-migrator.sh:29]: Can you check with the salt squad?

I am pretty sure they had an obsolete for `python2-salt` package at Salt 3002.

**Comment by mcalmer** [comment]
> > Afaik python2-salt does not exist anymore. We should try to add a weakremover in the release package
> 
> What is "a weakremover in the release package" ?

https://build.suse.de/package/view_file/SUSE:SLE-15-SP3:GA/000release-packages/weakremovers.inc?expand=1

I think this works for every SUSE product. Also for Leap....

It just tell libzypp, that it is ok to remove a package when there is nothing else which require it

**Comment by jordimassaguerpla** [comment]
Never used this before. I just branched all packages from systemsmanagement;Uyuni:Master into a subproject under my home project. Once builds finish, I will give it a try. python2-salt is required by the patterns-suma-server, so let's see what happens with the weak remove.

**Comment by mcalmer** [comment]
> python2-salt is required by the patterns-suma-server

this must go away. The package does not exist anymore.

**Comment by juliogonzalez** [comment]
suma-serverf!?

> > python2-salt is required by the patterns-suma-server
> 
> this must go away. The package does not exist anymore.

That package is gone, 14 days ago already.

But what worries me more is `required by the patterns-suma-server`???? That's a SUMA package. The pattern for the Uyuni server is `patterns-uyuni_server`

**Comment by jordimassaguerpla** [comment]
> suma-serverf!?
> 
> > > python2-salt is required by the patterns-suma-server
> > 
> > 
> > this must go away. The package does not exist anymore.
> 
> That package is gone, 14 days ago already.
> 
> But what worries me more is `required by the patterns-suma-server`???? That's a SUMA package. The pattern for the Uyuni server is `patterns-uyuni_server`

Sorry, I wrote it from memory, sure it is the uyuni-server one.

**Comment by jordimassaguerpla** [comment]
The issue is that salt does not obsolete python2-salt. The fix is in its way:

https://build.suse.de/request/show/241111

File: susemanager/susemanager.changes [hunk 2]
```diff
@@ -1,3 +1,4 @@
+- Update migration script to openSUSE 15.3
 -------------------------------------------------------------------
```
Comment (mcalmer) [inline susemanager/susemanager.changes:2]: better add a newline between your message and the separator.

**Review by mcalmer — APPROVED** [review]
one nitpick in the changelog

**Comment by jordimassaguerpla** [comment]
> The issue is that salt does not obsolete python2-salt. The fix is in its way:
> 
> https://build.suse.de/request/show/241111

I just did a smoke test by adding the repo in the request and I can confirm the problem with python2-salt goes away.

**Comment by jordimassaguerpla** [comment]
@juliogonzalez what about the postgresql conflict?

```
2 Problems:
Problem: nothing provides postgresql12-server = 12.6-3.24.1 needed by postgresql12-llvmjit-12.6-3.24.1.x86_64
Problem: nothing provides postgresql12-server = 12.6-3.24.1 needed by postgresql12-llvmjit-12.6-3.24.1.x86_64

Problem: nothing provides postgresql12-server = 12.6-3.24.1 needed by postgresql12-llvmjit-12.6-3.24.1.x86_64
 Solution 1: deinstallation of postgresql12-llvmjit-12.6-lp152.3.19.1.x86_64
 Solution 2: keep obsolete postgresql12-llvmjit-12.6-lp152.3.19.1.x86_64
 Solution 3: break postgresql12-llvmjit-12.6-3.24.1.x86_64 by ignoring some of its dependencies


```

File: susemanager/bin/server-migrator.sh (hunk 1 above)
Comment (jordimassaguerpla) [reply susemanager/bin/server-migrator.sh:29]: Debugged with Julio and found there is a pending release request that includes the obsoletes (see comment below)

**Comment by juliogonzalez** [comment]
In this case, I'd say an obsolete could be missing. Maybe it's on the queue (as the salt obsolete for python2-salt), I'd say you need to check with the Leap release engineers.

**Comment by jordimassaguerpla** [comment]
The reason why the issue with postgresql is not seen in sle but only in opensuse, is because postgresql12-server recommends the llvmjit subpackage only in openSUSE

**Comment by jordimassaguerpla** [comment]
This is a known issue

https://bugzilla.suse.com/show_bug.cgi?id=1183118
 Bug 1183118 - Upgrade to Leap 15.3 does not work for postgresql

The fix has been released yesterday to SLE-15-SP3.

I will ask about the release in leap.

**Comment by jordimassaguerpla** [comment]
We also need to add the updates coming from sle repo . This repo is set up on a leap 15.3 installation

**Comment by jordimassaguerpla** [comment]
Blocked on having these updates in leap 15.3:

https://build.suse.de/request/show/242901
https://build.suse.de/request/show/241111

**Comment by juliogonzalez** [comment]
> We also need to add the updates coming from sle repo . This repo is set up on a leap 15.3 installation

True!

**Comment by juliogonzalez** [comment]
@jordimassaguerpla is this ready? Why does it have WIP in the title?

File: susemanager/bin/server-migrator.sh [hunk 3]
```diff
@@ -22,18 +22,19 @@ echo
 spacewalk-service stop
 mv /etc/zypp/repos.d /etc/zypp/repos.d.old
```
Comment (juliogonzalez) [inline susemanager/bin/server-migrator.sh:23]: We should remove `/etc/zypp/repos.d.old/` first. People that migrated from 15.1 to 15.2 will already have it.

**Review by juliogonzalez — CHANGES_REQUESTED** [review]

File: susemanager/bin/server-migrator.sh (hunk 3 above)
Comment (juliogonzalez) [reply susemanager/bin/server-migrator.sh:23]:
```suggestion
rm -rf /etc/zypp/repos.d.old
mv /etc/zypp/repos.d /etc/zypp/repos.d.old
```

File: susemanager/bin/server-migrator.sh [hunk 4]
```diff
@@ -22,18 +22,19 @@ echo
...
-zypper ar -n "Main Repository" http://download.opensuse.org/distribution/leap/15.2/repo/oss repo-oss
-zypper ar -n "Main Update Repository" http://download.opensuse.org/update/leap/15.2/oss repo-update
-zypper ar -n "Non-OSS Repository" http://download.opensuse.org/distribution/leap/15.2/repo/non-oss repo-non-oss
-zypper ar -n "Update Repository (Non-Oss)" http://download.opensuse.org/update/leap/15.2/non-oss/ repo-update-non-oss
+zypper ar -n "Main Repository" http://download.opensuse.org/distribution/leap/15.3/repo/oss repo-oss
+zypper ar -n "Main Update Repository" http://download.opensuse.org/update/leap/15.3/oss repo-update
+zypper ar -n "Non-OSS Repository" http://download.opensuse.org/distribution/leap/15.3/repo/non-oss repo-non-oss
+zypper ar -n "Update Repository (Non-Oss)" http://download.opensuse.org/update/leap/15.3/non-oss/ repo-update-non-oss
 zypper ar -n "Uyuni Server Stable" https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Stable/images/repo/Uyuni-Server-POOL-x86_64-Media1/ uyuni-server-stable
+zypper ar -n "Update repository wiht updates from SUSE Linux Enterprise 15" http://download.opensuse.org/update/leap/15.3/sle repo-sle-update
```
Comment (juliogonzalez) [inline susemanager/bin/server-migrator.sh:30]:
```suggestion
zypper ar -n "Update repository wiht updates from SUSE Linux Enterprise 15" http://download.opensuse.org/update/leap/15.3/sle repo-sle-update
zypper ar -n "Update repository of openSUSE Backports" http://download.opensuse.org/update/leap/15.3/backports/
```

**Review by juliogonzalez — CHANGES_REQUESTED** [review]

**Review by juliogonzalez — APPROVED** [review]

**Pull request merged**